src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -51,31 +65,94 @@ data ArithmeticCircuit a = ArithmeticCircuit
         -- ^ The order of variable assignments
         acRNG      :: StdGen
     } deriving (Generic, NFData)
+--}
+
+data Circuit a = Circuit
+    {
+        acSystem   :: Map Natural (Constraint a),
+        -- ^ The system of polynomial constraints
+        acInput    :: [Natural],
+        -- ^ The input variables
+        acWitness  :: Map Natural a -> Map Natural a,
+        -- ^ The witness generation function
+        acVarOrder :: Map (Natural, Natural) Natural,
+        -- ^ The order of variable assignments
+        acRNG      :: StdGen
+    } deriving (Generic, NFData)
+
+data ArithmeticCircuit n a = ArithmeticCircuit
+  { acCircuit :: Circuit a
+  , acOutput  :: Vector n Natural
+  } deriving (Generic, NFData)
```

Comment:
This PR focuses on performance. I separated circuits and outputs so that when evaluating circuits, we can supply one circuit and only change output variables. See instance SymbolicData a (Vector n a), for example